proposals/0281-visible-forall.rst

```diff
@@ -623,8 +747,26 @@ Syntax
    This change applies to ``∀`` (the ``UnicodeSyntax`` rendition of ``forall``)
    as well.
 
-4. When ``ViewPatterns`` are enabled, interpret ``f (a -> b) = ...``
-   as a view pattern, otherwise as ``f ((->) a b) = ...``.
+4. Adding ``p1 -> p2`` to the syntax of patterns (see the BNF of "types-in-terms"
+   above) is incompatible with ``ViewPatterns``. When both ``RequiredTypeArguments``
+   and ``ViewPatterns`` are enabled, the conflict is resolved in favor of ``ViewPatterns``.
+   The programmer can write ``type (t1 -> t2)`` or ``(->) t1 t2`` instead.
```

Comment:
I think I'd favor making this treatment not dependent on `ViewPatterns`. As written, this proposal makes `ViewPatterns` fork-like, in that enabling `ViewPatterns` rejects some previously-valid programs. It is easy to avoid this, by unconditionally refusing to accept `p1 -> p2` as a function type pattern. It's a wart, yes, but I'd rather have a consistent wart than an inconsistent one.

(Really, the mistake here was allowing `ViewPatterns` not to have a keyword herald. This decision could only have happened because the implementation uses the same grammar for expressions and patterns. Just reading the Report, it would look impossible to parse a view pattern as implemented today, and a keyword herald would have been used. This might be a small lesson for the future.)

Reply:
Just to be clear, consider
```
data T a where
  K :: forall a -> a -> T a

f (K (Int->Int) x) = ...
```

* Richard proposes that the type pattern `(Int -> Int)` is rejected regardless of `ViewPatterns` because `Int -> Int` could be a view pattern.  
* Vlad suggests accepting it iff `ViewPattens` is off.

I don't feel strongly either way, but I'd be content to have it always-rejected.  You'd have to write
```
f (K ((->) Int Int) x) = ...
```
instead, which works regardless of `ViewPatterns`.

Reply:
> this proposal makes `ViewPatterns` fork-like, in that enabling `ViewPatterns` rejects some previously-valid programs

True, and I agree it's a problem. However, we only have that problem because `ViewPatterns` stole syntax that it had no business stealing, and we're now dealing with the consequences. So the question is: what's the best way to solve this problem?

Instead of limiting `RequiredTypeArguments` as you suggest, I'd rather have an alternative syntax for `ViewPatterns` that does not have this issue in the first place. There's already a draft (not mine) at https://gitlab.haskell.org/ghc/ghc/-/wikis/view-patterns-alternative, and I think it's quite promising.

Reply:
> problem because ViewPatterns

`ViewPatterns` has been around for over a decade (it is in GHC-7.0, I don't know when it was added). I'd consider it as one of the most stable extensions GHC has. The existence of `ViewPatternsAlternative` (wiki page created 12 years ago as well), and as far as I can tell zero progress on it is the same as *it doesn't exist*. *Someone* has to put work into advancing it first. Until then `ViewPatterns` should continue to work. `PatternSynonyms` usage relies on `ViewPatterns`.

I actually hope that GHC steering committee finds a way to avoid stalemates. alternative view patterns wiki page  has existed for 12 years, but there are no visible progress. But just because wiki page exists, `ViewPatterns` is not added to GHC2021 nor GHC2024, because "something might happen", but evidence says contrary. (FWIW, similarly "Safe Haskell" might improve, and not be in conflict with GHC202x, but so far there have been zero progress as well). That is uncomfortable situation. cc @adamgundry 
 @nomeata.

Reply:
> Someone has to put work into advancing it first.

This is exactly what's happening here. I'm trying to advance the "view patterns alternative" by providing a motivation for it. The syntax chosen for `ViewPatterns` has led to obstacles for `RequiredTypeArguments`, and I'm not willing to compromise `RequiredTypeArguments` because of that. This motivates the "view patterns alternative".

Motivation precedes implementation.

> Until then `ViewPatterns` should continue to work. `PatternSynonyms` usage relies on `ViewPatterns`.

Quite right, and no one is proposing to break `ViewPatterns` until we have the alternative. Is that your concern? You'll be happy to find that the amendment already says "the conflict is resolved in favor of `ViewPatterns`".

Reply:
Yes, the question here is not of breakage, but of "fork-like-ness" vs "accumulated defects".

I do think "fork-like-ness" is preferable too. As discussed elsewhere, the thing people don't like is breakage when their software is updated. Supporting multiple "contradictory" things for an extended period of time however is quite popular: maintainers of old code and writers of new code both get what they want.

Reply:
It seems to me there's a straightforward answer here: don't accept `Int -> Int` as a type pattern. If we end up doing an alternative syntax for view patterns, we can always add `Int -> Int` as a type pattern later. But if we accept `Int -> Int` as a type pattern now, we'll never be able to remove that, even if view patterns never get revised. I agree that this would be allowing a mistake around view patterns to make a new feature worse -- but I also think it's better to get 98% of the new feature in rather than wait to figure out view patterns before making any progress.

Reply:
I strongly agree with @goldfirere, regardless of the specific issue at hand. As a general principle, if there is a choice between options A and B, both improvements, and A is significantly easier to enact than B yet doesn't make it more difficult to enact B later, then A should be chosen. It's important to unlock value as quickly as possible, even (or maybe especially) partial value!

Reply:
@goldfirere I'm not looking forward to dropping this from the specification only to resubmit it again later. Would it be fine if I said that this change is conditional on having an alternative to view patterns?

Reply:
I pushed a change, adding this sentence

> When only ``RequiredTypeArguments`` is enabled, the conflict is not resolved either way. It is reported to the user in an error message. This can be reconsidered if an alternative syntax for view patterns is implemented in GHC.

Reply:
Yes I'm quite happy with that.